src/zelos/ext/plugins/snapshot/snapshot.py

```diff
@@ -0,0 +1,205 @@
+# Copyright (C) 2020 Zeropoint Dynamics
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+# ======================================================================
+import base64
+import json
+import logging
+import os
+
+from termcolor import colored
+
+from zelos import CommandLineOption, IPlugin, Zelos
+
+
+CommandLineOption(
+    "snapshot", action="store_true", help="Output a snapshot of memory."
+)
+
+
+class Snapshotter(IPlugin):
+    """
+    Output a snapshot of memory regions at the end of emulation.
+    """
+
+    def __init__(self, z: Zelos):
+        super().__init__(z)
+
+        self.logger = logging.getLogger(__name__)
+
+        if z.config.snapshot:
+            if z.config.verbosity == 0:
+                self.logger.error(
+                    (
+                        f"You will not get instruction comments without "
+                        f'running in verbose mode. Include this flag ("-vv") '
+                        f"if you want instruction comments in your snapshot. "
+                        f"For an additional speedup, consider also including "
+                        f'the fasttrace flag ("-vv --fasttrace").'
+                    )
+                )
+            original_file_name = z.internal_engine.original_file_name
+
+            def closure():
+                self.snapshot(original_file_name + ".zmu")
+                self.logger.info(
+                    f"Wrote snaphot to: "
+                    f"{os.path.abspath(original_file_name)}.zmu"
+                )
+
+            self.zelos.hook_close(closure)
+
+    def _bad_section(self, data):
+        # if size is too large, this is a bad section
+        if len(data) > 0x100000000:
+            self.logger.info(f"Data too large: 0x{len(data):x}")
+            return True
+
+        # if the data contains mostly zeros, we can ignore it
+        def percent_zeros(data):
+            num = 0
+            for c in data:
+                if c == 0:
+                    num += 1
+            return num / (1.0 * len(data))
+
+        pct_zeros = percent_zeros(data)
+        if pct_zeros > 0.999999:
+            self.logger.info(f"Mostly zeros, pct: {pct_zeros}")
+            return True
+
+        return False
+
+    def snapshot(self, outfile="emu.out"):
+        z = self.zelos.internal_engine
```

Comment:
We shouldn't refer to internal_engine as z anymore, we should move to only using z to refer to `Zelos`